controllers/devworkspace/solver/endpoint_strategy.go

```diff
@@ -72,8 +72,10 @@ func (u UsernameWkspName) getEndpointPathPrefix(endpointPath string) string {
 func (u UsernameWkspName) getHostname(endpointInfo *EndpointInfo, baseDomain string) string {
 	subDomain := fmt.Sprintf("%s-%s-%s", u.username, u.workspaceName, endpointInfo.endpointName)
 	if errs := validation.IsValidLabelValue(subDomain); len(errs) > 0 {
-		// if subdomain is not valid, use legacy paths
-		return fmt.Sprintf("%s-%d.%s", u.workspaceID, endpointInfo.order, baseDomain)
+		// If subdomain is not valid (e.g. too long), use alternate format
+		// The below should always be under 63 characters, as endpoint names are limited to 15 characters and workspace IDs are
+		// 25 characters.
+		return fmt.Sprintf("%s-%s.%s", u.workspaceID, endpointInfo.endpointName, baseDomain)
```

Comment:
What about sorting endpoints by theirs names here [1] instead of changing host names?
I just have some concerns, maybe someone relies on hostnames. 

[1] https://github.com/eclipse-che/che-operator/blob/7c7bd4f796eb083a21cf5ba5f750cde1d69e901a/controllers/devworkspace/solver/che_routing.go#L485-L486

Reply:
I considered sorting the endpoints instead of this approach, but honestly iterating through a map in sorted order in Go is annoying (since there's no way to easily get a keyset, etc.) :) 

These hostnames are not going to be reliably static in any case since this functionality is triggered only when the default endpoint name becomes too long. For example, some users might see the normal format while others see the legacy format (depending on how long usernames are). You could even end up in a situation where half your endpoints are normal, and half are legacy. In the past, we've pointed people at the `endpoint-url` attribute to determine the actual hostname used for each devfile endpoint.

I'm fine with sorting the endpoints, though I had a small concern that I was missing something in how they're handled, as the [original issue](https://github.com/eclipse/che/issues/22667) has `order` numbers going up to 13 (which may just be how many endpoints the devfile has, I don't know). 

An additional bonus of doing it this way is that we don't run into situations where endpoints could be updated unexpectedly -- e.g. with order, even sorted, if you update the devfile to have an endpoint earlier in the order, all your legacy-hostname endpoints might change.

Reply:
Thank you for explanation, make sense to me.